### DISCUSSION TITLE
this.accessors.getLocationCentroid is not a function     at FlowMapAggregateAccessors.getLocationCentroid (FlowMapAggregateAccessors.js?3933:20:1)

### DISCUSSION THREAD

**Discussion by mrosminin** [visgl/flowmap.gl#26] [Q&A] [answered]
Hi,

Trying to use FlowMapLayer in a vue.js app. Getting error 
this.accessors.getLocationCentroid is not a function
    at FlowMapAggregateAccessors.getLocationCentroid (FlowMapAggregateAccessors.js?3933:20:1)

data is FlowMapData

![image](https://user-images.githubusercontent.com/46540893/233962533-fab203c1-cc72-4620-ba8d-fa820db187ff.png)

![image](https://user-images.githubusercontent.com/46540893/233962572-8955a62c-eef6-48b3-a527-c40d19bc814a.png)

Adding layer exactly as in example https://github.com/visgl/flowmap.gl/blob/main/examples/webpack-app/src/app.js

![image](https://user-images.githubusercontent.com/46540893/233963085-699f3243-002a-42e2-8f69-a5597c039106.png)

what ca be the problem with this?

**Comment by ilyabo** [comment, marked as answer]
`getLocationCentroid` was refactored into `getLocationLat/Lon`. Check the installed dependency versions. Maybe there's a mismatch between `data` and `layers`. They should be the same:
``` 
    "@flowmap.gl/data": "8.0.0-alpha.25",
    "@flowmap.gl/layers": "8.0.0-alpha.25",
```

**Comment by mrosminin** [comment]
Thanks, this was the metter.

One more question. For what reason clustering may not occur?


https://user-images.githubusercontent.com/46540893/234182615-0d7b9ad8-a3ca-401a-a5ff-8f2adfc30fdf.mov

**Comment by ilyabo** [comment]
From your screenshot it looks like you are always passing `clusteringEnabled: false`

**Comment by mrosminin** [comment]
Of course true, checked twice! Ive tried on another dataset, and it's clustering but only for 1 level.
If clusteringAuto: true, only for 1 level on "small" dataset, the "large" dataset above - no clustering at all.

https://user-images.githubusercontent.com/46540893/234248230-fab03e72-13cf-4ee5-8064-7069a7c33623.mov

If setting clusteringAuto false and clusteringLevel unlike 1, all objects and flows disappear from map.

**Reply by ilyabo** [reply]
Are you using HCA or H3? 
Not sure what it could be. You might just need to zoom out farther. When using the clusteringLevel slider, some levels might be empty if the clusters are the same as on a lower zoom level.

**Comment by mrosminin** [comment]
[flows.txt](https://github.com/visgl/flowmap.gl/files/11322600/flows.txt)
[locations.txt](https://github.com/visgl/flowmap.gl/files/11322601/locations.txt)

No matter h3 or HCA. On this dataset no clustering.


config
{
    "darkMode": false,
    "colorScheme": "Blues",
    "highlightColor": "#ff9b29",
    "opacity": 0.5,
    "fadeEnabled": true,
    "fadeOpacityEnabled": false,
    "fadeAmount": 50,
    "clusteringEnabled": true,
    "clusteringAuto": true,
    "clusteringLevel": 5,
    "clusteringMethod": "HCA",
    "animationEnabled": false,
    "adaptiveScalesEnabled": true,
    "locationTotalsEnabled": true,
    "locationLabelsEnabled": false,
    "maxTopFlowsDisplayNum": 5000
}

**Reply by ilyabo** [reply]
Looks like the clustering algorithm has issues with the fact that many of the locations with different IDs share exactly the same coordinates. I will try to fix the issue in the library. In the meantime, you could try to modify your dataset to make sure same locations have the same IDs. Or add a small random displacement to the coordinates.

**Comment by mrosminin** [comment]
Yes, this was the matter. I modified dataset, and it started to work properly.
